packages/sanity/src/core/search/common/deriveSearchWeightsFromType.ts

```diff
@@ -0,0 +1,237 @@
+import {type CrossDatasetType, type SchemaType, type SearchConfiguration} from '@sanity/types'
+
+import {isRecord} from '../../util'
+import {type SearchPath, type SearchSpec} from './types'
+
+interface SearchWeightEntry {
+  path: string
+  weight: number
+  type: 'string' | 'pt'
+}
+
+const CACHE = new WeakMap<SchemaType | CrossDatasetType, SearchSpec>()
+const PREVIEW_FIELD_WEIGHT_MAP = {
+  title: 10,
+  subtitle: 5,
+  description: 1.5,
+}
+const BASE_WEIGHTS: Record<string, Omit<SearchWeightEntry, 'path'>> = {
+  _id: {weight: 1, type: 'string'},
+  _type: {weight: 1, type: 'string'},
+}
+const GROQ_RESERVED_KEYWORDS = ['true', 'false', 'null']
+const builtInObjectTypes = ['reference', 'crossDatasetReference']
+
+const getTypeChain = (type: SchemaType | undefined): SchemaType[] =>
+  type ? [type, ...getTypeChain(type.type)] : []
+
+const isPtField = (type: SchemaType | undefined) =>
+  type?.jsonType === 'array' &&
+  type.of.some((arrType) => getTypeChain(arrType).some(({name}) => name === 'block'))
+
+const isStringField = (schemaType: SchemaType | undefined) =>
+  getTypeChain(schemaType).some((type) => type.name === 'string')
+
+const isSearchConfiguration = (options: unknown): options is SearchConfiguration =>
+  isRecord(options) && 'search' in options && isRecord(options.search)
+
+function isSchemaType(input: SchemaType | CrossDatasetType | undefined): input is SchemaType {
+  return typeof input !== 'undefined' && 'name' in input
+}
+
+/**
+ * Serialize field path for GROQ query.
+ *
+ * Reserved keywords, such as `null`, cannot be accessed using dot notation. These fields will
+ * instead be serialized using square bracket notation.
+ */
+function pathToString(...path: string[]): string {
```

Comment:
It's a good point, and crossed my mind when I added it. I landed on this implementation because it works for our current use case without having any wider impact. Now things are working, I'll see if we can refactor.

Reply:
Refactored in #5987.